### ISSUE TITLE
spec: GrantRoundPayout contract

### ISSUE THREAD

**Issue by thelostone-mc** [dcgtc/dgrants#79]
### Description

Once a GrantRound has ended, the `payoutAdmin` would invoke [payoutGrants](https://github.com/dcgtc/dgrants/blob/main/contracts/contracts/GrantRound.sol#L98) to transfer funds into the `GrantRoundPayout` contract.

### What is this contract do 

**Global variables**
- `mapping(address => uint256) public payouts;`  stores the address to amount 

**Functions** 
- invoke function `setPayout` to be able to upload the distribution  of type `PayoutFields` (this function should be callable multiple times if incase all payouts can't be set in 1 transaction  )
- invoke function `payoutGrants` which 
  - checks to ensure total amount in `payouts` =< amount in contract 
  - loop through the grants and do the actual transfer 

```
mapping(address => uint256) public payouts;

struct PayoutFields {
  address recipient; // grant receiving address
  uint256 amount; // match amount for that grant
}
```

**Comment by thelostone-mc** [comment]
Another way which is cheaper for payouts is where we expect users to come and claim payouts and offloading the gas onto them instead. While that is better -> maybe we reserve this for v2 version

**Comment by apbendi** [comment]
Some thoughts

* Instead of having loops in the methods themselves, we can have `setPayout` and `claimPayout` functions that take a single address, then add a [multicall method](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/v4.2.0/contracts/utils/Multicall.sol) for bulk operations
* There needs to be an `owner` address and `setPayout` needs to be `onlyOwner`
* There should also be a `finalizePayout` method that is `onlyOwner` and flips a switch that prevents future edits to the payouts
* There should be a `payoutToken` immutable that stores the address of the payout token and is set in the constructor
* Anyone should be able to call the `claimPayout` method, pass an address, and initiate the transfer on behalf of the payee. This means we could have grant owners claim themselves, or execute the payouts ourselves.
* Question: Instead of a payee address, should we instead have a reference to the Registry and a GrantId to store the payout mapping? The thought here is that the payee address can change, so we should treat the GrantId as the canonical source of "how to pay each Grant". The `claimPayout` method would look up the payee address at the time of claim before doing the transfer. Downside here is extra gas, though this could be optimized later via #54

**Comment by mds1** [comment]
@thelostone-mc You are basically describing using the current [`MatchPayouts`](https://github.com/gitcoinco/matching_contracts/blob/main/contracts/MatchPayouts.sol) contract right?

We wrote that contract as a quick solution with plans to upgrade to a Merkle distributor, and here I would just recommend jumping straight to the Merkle distributor. You can argue that it will be overkill for the MVP as there will only be a few grants, but:
1. It's actually not that hard to implement, and then it's "done", i.e. we don't need to worry about upgrading that after the MVP
2. It solves many of the above comments from @apbendi, i.e. we don't need loops or multicall methods, no "out of gas" concerns, no `owner` is needed
3. Much cheaper to set the distribution on a Merkle distributor than it is with a mapping, especially as the number of grants grows

I've written a Merkle Distributor recently just to get familiar with the process, and it's pretty simple by basing it off Uniswap's [implementation](https://github.com/Uniswap/merkle-distributor)

------------------------

If we don't go with a Merkle distributor, I agree with all of @apbendi's comments. I don't feel strongly about whether we use the payee address or grant ID, because it will be obsoleted by https://github.com/dcgtc/dgrants/discussions/54 which I personally think should be implemented

**Comment by apbendi** [comment]
Hey @thelostone-mc and @gdixon, what do you think of @mds1 argument that a Merkle Distributor isn't much harder to implement than a naive storage solution, given the Uniswap implementation exists? I think he might have convinced me. One thing to consider: we'll have to make the payout script generate a Merkle root in addition to the list of payouts. Seems like the Uniswap code makes that easy, but want to see what you think?

**Comment by gdixon** [comment]
@apbendi - following the discussion on Friday, I think this makes a lot of sense. I will have a look at the Uniswap Merkle implementation and see how we can go about connecting everything in. @thelostone-mc we can likely just write `hash/merkle.ts` and set up a separate script to generate the proofs 🙌

**Comment by mds1** [comment]
@gdixon For some references:
- [Here](https://github.com/Uniswap/merkle-distributor) is Uniswap's implementation that became a common template. You'll notice their `claim` method takes an index
- [Here](https://github.com/mds1/auction-house) is an implementation I did recently based off that. I modified Uniswap's `balance-tree.ts` into `split-tree`.ts` so the tree is based on percentages of some arbitrary value in the contract instead of amounts, but either approach works. I also modified it so the contract's `claim` method doesn't need an index anymore so it's a bit more user friendly (I'm not sure why they used the `index`, so might be worth sanity checking that decision to remove it)

The README + tests for each of those should demonstrate the full flow of generating the root and making claims, but let me know if you have any questions!
